File: javascript/queen-attack/queen-attack.js
function Queens(input){
    if (input) {
        if (input.white[0] === input.black[0] && input.white[1] === input.black[1]) {
            throw 'Queens cannot share the same space';
        }
        this.white = input.white;
        this.black = input.black;
    } else {
        this.white = [0, 3];
        this.black = [7, 3];
    }
}

Queens.prototype.toString = function(){
    var board = getEmptyBoard();
    board[this.white[0]][this.white[1]] = 'W';
    board[this.black[0]][this.black[1]] = 'B';
    board = board.map(function(line){
        return line.join(' ');
    });
    return board.join('\n') + '\n';
}

Queens.prototype.canAttack = function(){
    // row or column
    if(this.white[0] === this.black[0] || this.white[1] === this.black[1]) return true;
    // diagonal
    if(Math.abs(this.white[0]-this.black[0]) === Math.abs(this.white[1]-this.black[1])) return true;
    return false;
}

function getEmptyBoard(){
    var array = [];
    for(var i = 0; i < 8; i++) {
        array.push(Array(8).fill('_'));
    }
    return array;
}

module.exports = Queens;

File: go/queen-attack/queen_attack.go
package queenattack

import (
	"errors"
)

const testVersion = 2

// CanQueenAttack checks whether the queens on the given fields can attack each other
func CanQueenAttack(white string, black string) (bool, error) {
	if !valid(white) || !valid(black) {
		return false, errors.New("invalid input")
	}
	// same position
	if white == black {
		return false, errors.New("same position")
	}
	// same row or column
	if white[0] == black[0] || white[1] == black[1] {
		return true, nil
	}
	// same diagonal
	if distance(white[0], black[0]) == distance(white[1], black[1]) {
		return true, nil
	}
	return false, nil
}

func valid(pos string) bool {
	if len(pos) != 2 {
		return false
	}
	if pos[0] < 'a' || pos[0] > 'h' {
		return false
	}
	if pos[1] < '1' || pos[1] > '8' {
		return false
	}
	return true
}

func distance(a, b byte) byte {
	if a < b {
		a, b = b, a
	}
	return a - b
}
